code/game/machinery/computer/apc_control.dm

```diff
@@ -0,0 +1,208 @@
+			dat += "<b>Name:</b> <a href='?src=\ref[src];name_filter=1'>[filters["Name"] ? filters["Name"] : "None set"]</a><br>"
+			dat += "<b>Charge:</b> <a href='?src=\ref[src];above_filter=1'>\>[filters["Charge Above"] ? filters["Charge Above"] : "NaN"]%</a> and <a href='?src=\ref[src];below_filter=1'>\<[filters["Charge Below"] ? filters["Charge Below"] : "NaN"]%</a><br>"
+			dat += "<b>Accessible:</b> <a href='?src=\ref[src];access_filter=1'>[filters["Responsive"] ? "Non-Responsive Only" : "All"]</a><br><br>"
+			for(var/A in apcs)
+				var/obj/machinery/power/apc/APC = apcs[A]
+				if(filters["Name"] && !findtext(APC.name, filters["Name"]) && !findtext(APC.area.name, filters["Name"]))
+					continue
+				if(filters["Charge Above"] && (APC.cell.charge / APC.cell.maxcharge) < filters["Charge Above"] / 100)
+					continue
+				if(filters["Charge Below"] && (APC.cell.charge / APC.cell.maxcharge) > filters["Charge Below"] / 100)
+					continue
+				if(filters["Responsive"] && !APC.aidisabled)
+					continue
+				dat += "<a href='?src=\ref[src];access_apc=\ref[APC]'>[A]</a><br>\
+				<b>Charge:</b> [APC.cell.charge] / [APC.cell.maxcharge] W ([round((APC.cell.charge / APC.cell.maxcharge) * 100)]%)<br>\
+				<b>Area:</b> [APC.area]<br>\
+				[APC.aidisabled || APC.panel_open ? "<font color='#FF0000'>APC does not respond to interface query.</font>" : "<font color='#00FF00'>APC responds to interface query.</font>"]<br><br>"
+			dat += "<a href='?src=\ref[src];check_logs=1'>Check Logs</a><br>"
+			dat += "<a href='?src=\ref[src];log_out=1'>Log Out</a><br>"
+			if(emagged)
+				dat += "<font color='#FF0000'>WARNING: Logging functionality partially disabled from outside source.</font><br>"
+				dat += "<a href='?src=\ref[src];restore_logging=1'>Restore logging functionality?</a><br>"
+		else
+			if(logs.len)
+				for(var/entry in logs)
+					dat += "[entry]<br>"
+			else
+				dat += "<i>No activity has been recorded at this time.</i><br>"
+			if(emagged)
+				dat += "<a href='?src=\ref[src];clear_logs=1'><font color='#FF0000'>@#%! CLEAR LOGS</a>"
+			dat += "<a href='?src=\ref[src];check_apcs=1'>Return</a>"
+		operator = user
+	else
+		dat = "<a href='?src=\ref[src];authenticate=1'>Please swipe a valid ID to log in...</a>"
+	var/datum/browser/popup = new(user, "apc_control", name, 600, 400)
+	popup.set_content(dat)
+	popup.set_title_image(user.browse_rsc_icon(icon, icon_state))
+	popup.open()
+
+/obj/machinery/computer/apc_control/Topic(href, href_list)
+	if(..())
+		return
+	var/image/I = image(src) //For feedback message flavor
+	if(href_list["authenticate"])
+		var/obj/item/weapon/card/id/ID = usr.get_active_held_item()
+		if(!istype(ID))
+			ID = usr.get_idcard()
+		if(ID && istype(ID))
+			if(check_access(ID))
+				authenticated = TRUE
+				auth_id = "[ID.registered_name] ([ID.assignment])"
+				log_activity("logged in")
+		if(!authenticated) //Check for emags
+			var/obj/item/weapon/card/emag/E = usr.get_active_held_item()
+			if(E && istype(E))
```

Comment:
you cannot trust the topic call to do that, someone can forge the call and hold the emag item in their hand